unstable/lens/isomorphism/base.rkt

```diff
@@ -0,0 +1,67 @@
+#lang racket/base
+
+(provide isomorphism-lens
+         isomorphism-lens?
+         isomorphism-lens-inverse
+         isomorphism-lenses
+         )
+(module+ data
+  (provide string->symbol-lens symbol->string-lens
+           number->string-lens string->number-lens
+           list->vector-lens vector->list-lens
+           list->string-lens string->list-lens
+           hash->list-lens list->hash-lens
+           ))
+
+(require racket/match
+         lens/base/main
+         )
+(module+ test
+  (require rackunit (submod ".." data)))
+
+(struct isomorphism-lens (f inv) #:transparent
+  #:methods gen:lens
+  [(define (lens-view lens tgt)
+     ((isomorphism-lens-f lens) tgt))
+   (define (lens-set lens tgt v)
+     ((isomorphism-lens-inv lens) v))])
+
+(define (isomorphism-lens-inverse lens)
+  (match lens
+    [(isomorphism-lens f inv)
+     (isomorphism-lens inv f)]))
+
+(define (isomorphism-lenses f inv)
+  (values (isomorphism-lens f inv)
+          (isomorphism-lens inv f)))
+
+(module+ data
+  (define-values [string->symbol-lens symbol->string-lens]
+    (isomorphism-lenses string->symbol symbol->string))
+  (define-values [number->string-lens string->number-lens]
+    (isomorphism-lenses number->string string->number))
+  (define-values [list->vector-lens vector->list-lens]
+    (isomorphism-lenses list->vector vector->list))
+  (define-values [list->string-lens string->list-lens]
+    (isomorphism-lenses list->string string->list))
+  (define-values [hash->list-lens list->hash-lens]
+    (isomorphism-lenses hash->list make-immutable-hash))
```

Comment:
I just ran this and came across a case that should have been obvious: when two keys are the same:
```racket
#lang racket/base
(require rackunit racket/contract)
(test-begin
 (for ([i (in-range 1000)])
   (define lst (contract-random-generate (listof pair?)))
   (check-equal? (hash->list (make-immutable-hash lst)) lst)))
```
This failed with:
```
--------------------
FAILURE
actual:     ((1 . #<procedure:random-any/c-generated-procedure>) (0 . #<procedure:random-any/c-generated-procedure>) (1/3 . #<procedure:random-any/c-generated-procedure>) (#"\364" . #\ᓛ))
expected:   ((1/3 . #<procedure:random-any/c-generated-procedure>) (0 . #<procedure:random-any/c-generated-procedure>) (1 . 4) (#"\364" . #\ᓛ) (1 . #<procedure:random-any/c-generated-procedure>))
name:       check-equal?
location:   (unsaved-editor 6 3 155 57)
expression: (check-equal? (hash->list (make-immutable-hash lst)) lst)

. Check failure
--------------------
```
Because there were two pairs in the original list with the key `1`, and only one of those stayed in the hash table. The order is also moved around, so there's that too.